File: repos/koi-web/src/lib/repositories/message_repository.ts
import { db } from '../db';
import { cryptoService } from '../crypto';
import type { Message } from '../../models/index';

export class MessageRepository {
  async save(message: Message): Promise<void> {
    const toStore: Message = {
      ...message,
      content: await cryptoService.encrypt(message.content),
    };
    await db.messages.put(toStore);
  }

  async getById(id: string): Promise<Message | undefined> {
    const msg = await db.messages.get(id);
    if (!msg) return undefined;
    return { ...msg, content: await cryptoService.decrypt(msg.content) };
  }

  async getByConversation(conversationId: string): Promise<Message[]> {
    const msgs = await db.messages
      .where('conversationId')
      .equals(conversationId)
      .sortBy('timestamp');
    return Promise.all(
      msgs.map(async (m) => ({ ...m, content: await cryptoService.decrypt(m.content) })),
    );
  }

  async getUnsynced(): Promise<Message[]> {
    const msgs = await db.messages.where('synced').equals(0).toArray();
    return Promise.all(
      msgs.map(async (m) => ({ ...m, content: await cryptoService.decrypt(m.content) })),
    );
  }

  async markSynced(id: string): Promise<void> {
    await db.messages.update(id, { synced: true });
  }

  async markAllSynced(ids: string[]): Promise<void> {
    await db.transaction('rw', db.messages, async () => {
      for (const id of ids) {
        await db.messages.update(id, { synced: true });
      }
    });
  }

  async deleteByConversation(conversationId: string): Promise<void> {
    await db.messages.where('conversationId').equals(conversationId).delete();
  }

  async delete(id: string): Promise<void> {
    await db.messages.delete(id);
  }
}

export const messageRepository = new MessageRepository();

File: repos/koi-backend/flutter_lib/lib/storage/repositories/message_repository.dart
import 'package:hive/hive.dart';
import '../hive_manager.dart';
import '../../models/hive_message_model.dart';

class MessageRepository {
  Box<HiveMessage> get _box => HiveManager.messages;

  Future<void> saveMessage(HiveMessage message) async =>
      _box.put(message.id, message);

  Future<void> saveMessages(List<HiveMessage> messages) async {
    final map = {for (final m in messages) m.id: m};
    await _box.putAll(map);
  }

  HiveMessage? getMessage(String id) => _box.get(id);

  List<HiveMessage> getByConversation(String conversationId) {
    return _box.values
        .where((m) => m.conversationId == conversationId)
        .toList()
      ..sort((a, b) => a.timestamp.compareTo(b.timestamp));
  }

  List<HiveMessage> getUnsynced() =>
      _box.values.where((m) => !m.synced).toList();

  Future<void> markSynced(String id) async {
    final message = _box.get(id);
    if (message == null) return;
    message.synced = true;
    await message.save();
  }

  Future<void> markAllSynced(List<String> ids) async {
    for (final id in ids) {
      await markSynced(id);
    }
  }

  Future<void> deleteMessage(String id) async => _box.delete(id);

  Future<void> deleteConversation(String conversationId) async {
    final keys = _box.values
        .where((m) => m.conversationId == conversationId)
        .map((m) => m.id)
        .toList();
    await _box.deleteAll(keys);
  }
}
